syntax = "proto3";

package lms;

service LMS {
    rpc Register(RegisterRequest) returns (StatusResponse);
    rpc Login(LoginRequest) returns (LoginResponse);
    rpc Logout(LogoutRequest) returns (StatusResponse);
    rpc Post(PostRequest) returns (StatusResponse);
    rpc Get(GetRequest) returns (GetResponse);
}

message RegisterRequest {
    string username = 1;
    string password = 2;
    string role = 3;  // student or teacher
    string name = 4;  // Additional field for user name
}

message LoginRequest {
    string username = 1;
    string password = 2;
}

message LoginResponse {
    string status = 1;
    string token = 2;
    string role = 3;  // Include the role
}

message LogoutRequest {
    string token = 1;
}

message StatusResponse {
    string status = 1;
    string error_code = 2; // Optional field for error codes
}

// Main field types

message AssignmentData {
    string student_name = 1;
    string teacher_name = 2;
    string filename = 3;
    string file_path = 4;  // Path to the uploaded file
    string grade = 5;      // Field for grade
    string feedback_text = 6;  // Field for feedback
    string submission_date = 7;
    string assignment_id = 8;
}

message FeedbackData {
    string student_name = 1;
    string teacher_name = 2;  // Added the type
    string feedback_text = 3;  // Text of the feedback
    string submission_date = 4;
    string feedback_id = 5;
}

message CourseMaterial {
    string teacher_name = 1;
    string course_name = 2;
    string filename = 3;
    string file_path = 4;
    string material_id = 5;
    string upload_date = 6;
}


// Post requests
message PostRequest {
    string token = 1;
    oneof data_type {  // Define different data types that can be posted
        AssignmentData assignment = 2;
        FeedbackData student_feedback = 3;  // Updated to include feedback
        CourseContent content = 4;
    }
}


// Get requests
message GetRequest {
    string token = 1;
    oneof data_type {  // Define types of data that can be fetched
        AssignmentData assignment = 1;
        FeedbackData feedback = 2;
        CourseContent content = 3;
    }
}


message GetResponse {
    string status = 1;
    repeated AssignmentData assignment_items = 2;  // Existing data items for assignments and course materials
    repeated FeedbackData feedback_items = 3;  // New feedback items
    repeated CourseContent content_items = 4;
}


